plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/client/ElasticsearchClient.java

```diff
@@ -538,45 +623,166 @@ public boolean indexExists(String index)
+        return new IndexMetadata.ObjectType(fields);
+    }
+
+    private List<IndexMetadata.Field> merge(String index, String parentPrefix, List<IndexMetadata.Field> schemaFields1, List<IndexMetadata.Field> schemaFields2)
+    {
+        // Non object type fields
+        Set<IndexMetadata.Field> fields = Stream.concat(schemaFields1.stream(), schemaFields2.stream())
+                .filter(field -> !(field.getType() instanceof IndexMetadata.ObjectType))
+                .collect(toSet());
+
+        // Object Type fields
+        Set<IndexMetadata.Field> objectTypeFields = Stream.concat(schemaFields1.stream(), schemaFields2.stream())
+                .filter(field -> field.getType() instanceof IndexMetadata.ObjectType)
+                .collect(groupingBy(field -> field.getName()))
+                .entrySet()
+                .stream()
+                .map(fieldList -> mergeNestedFields(fieldList.getKey(), fieldList.getValue(), index, parentPrefix))
+                .collect(toSet());
+
+        fields.addAll(objectTypeFields);
+        fields = checkAndCorrectMismatchFields(index, parentPrefix, fields);
+
+        return ImmutableList.copyOf(fields);
+    }
+
+    private IndexMetadata.Field mergeNestedFields(String fieldName, List<IndexMetadata.Field> fields, String index, String parentPrefix)
+    {
+        String prefix = isNullOrEmpty(parentPrefix) ? fieldName : getFlattenedKey(parentPrefix, fieldName);
+        IndexMetadata.ObjectType type = fields.stream()
+                .map(field -> (IndexMetadata.ObjectType) field.getType())
+                .reduce((nestedSchema1, nestedSchema2) -> merge(index, prefix, nestedSchema1, nestedSchema2))
+                .get();
+        IndexMetadata.Field indexField = fields.iterator().next();
+        return new IndexMetadata.Field(indexField.isArray(), indexField.getName(), type);
+    }
+
+    private Set<IndexMetadata.Field> checkAndCorrectMismatchFields(String index, String parent, Set<IndexMetadata.Field> fields)
+    {
+        final Map<String, List<IndexMetadata.Field>> fieldsByName = fields.stream().collect(groupingBy(IndexMetadata.Field::getName));
+
+        final Set<IndexMetadata.Field> resultFields = fieldsByName.values().stream()
+                .filter(fieldList -> fieldList.size() == 1)
+                .map(fieldList -> fieldList.get(0))
+                .collect(toSet());
+
+        final Map<String, List<IndexMetadata.Type>> columnsWithMultipleDataTypes = fieldsByName.entrySet()
+                .stream()
+                .filter(fieldNameToFieldsList -> fieldNameToFieldsList.getValue().size() > 1)
+                .collect(toImmutableMap(Map.Entry::getKey, columnToFields -> columnToFields.getValue().stream()
+                        .map(IndexMetadata.Field::getType)
+                        .collect(toList())));
+
+        if (!columnsWithMultipleDataTypes.isEmpty() && failOnAliasSchemaMismatch) {
```

Comment:
If difference is `isArray` then we can't coerce right? Because `Array(String)` and `String` should not be coerced. In such cases, let users handle the mapping with uniform `isArray` property value across indices.

Reply:
since ES allows arrays for any type (and that we allow the array type using the specific isArray property), IMO array<T> would be the "encompassing" type, loosely speaking. As every element typed `T` can be easily transformed into `array<T>`, I'd just use the latter for coercing.  

Similarly asRawJson takes priority currently over defined types as seen in tests like `testAsRawJsonForAllPrimitiveTypes`. So I'd just use json in that case. 

FWIW, even if we end up just rejecting such cases (which I think we shouldn't), I think a more explicit error message would be very useful here as types match otherwise.

Reply:
We cast every such case to `rawJson`. If fields have different types/properties then we cast it as `rawJson`.

Reply:
Handled. 
If same field has different `isArray` with same type then cast as `array`.
If same field has different `asRawJson` with same type then cast as `rawJson`.